<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0321ENSkillsNetwork928-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="400" alt="Skills Network Logo"  />
    </a>
</p>


# **Data Wrangling Lab**


Estimated time needed: **45 to 60** minutes


In this assignment you will be performing data wrangling.


## Objectives


In this lab you will perform the following:


-   Identify duplicate values in the dataset.

-   Remove duplicate values from the dataset.

-   Identify missing values in the dataset.

-   Impute the missing values in the dataset.

-   Normalize data in the dataset.


<hr>


## Hands on Lab


Import pandas module.


In [1]:
import pandas as pd

<ipython-input-1-7dd3504c366f>:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Load the dataset into a dataframe.


<h2>Read Data</h2>
<p>
We utilize the <code>pandas.read_csv()</code> function for reading CSV files. However, in this version of the lab, which operates on JupyterLite, the dataset needs to be downloaded to the interface using the provided code below.
</p>


The functions below will download the dataset into your browser:


In [2]:
from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

In [3]:
file_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DA0321EN-SkillsNetwork/LargeData/m1_survey_data.csv"

To obtain the dataset, utilize the download() function as defined above:  


In [4]:
await download(file_path, "m1_survey_data.csv")
file_name="m1_survey_data.csv"

Utilize the Pandas method read_csv() to load the data into a dataframe.


In [5]:
df = pd.read_csv(file_name)

> Note: This version of the lab is working on JupyterLite, which requires the dataset to be downloaded to the interface.While working on the downloaded version of this notebook on their local machines(Jupyter Anaconda), the learners can simply **skip the steps above,** and simply use the URL directly in the `pandas.read_csv()` function. You can uncomment and run the statements in the cell below.


In [6]:
#df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DA0321EN-SkillsNetwork/LargeData/m1_survey_data.csv")

## Finding duplicates


In this section you will identify duplicate values in the dataset.


 Find how many duplicate rows exist in the dataframe.


In [10]:
# your code goes here

duplicate_count = df.duplicated().sum()

# Display the first few duplicate rows to understand their structure
duplicate_rows = df[df.duplicated()]

print(f"Number of duplicate rows: {duplicate_count}")
print("\nFirst few duplicate rows:")
print(duplicate_rows.head())

Number of duplicate rows: 154

First few duplicate rows:
      Respondent                      MainBranch Hobbyist  \
1168        2339  I am a developer by profession      Yes   
1169        2342  I am a developer by profession      Yes   
1170        2343  I am a developer by profession      Yes   
1171        2344  I am a developer by profession      Yes   
1172        2347  I am a developer by profession      Yes   

                                            OpenSourcer  \
1168                         Once a month or more often   
1169                                              Never   
1170  Less than once a month but more than once per ...   
1171                                              Never   
1172                                              Never   

                                             OpenSource          Employment  \
1168  OSS is, on average, of HIGHER quality than pro...  Employed full-time   
1169  The quality of OSS and closed source software ...  Employ

In [8]:
# additional code for Graded Quiz 
# How many duplicate values are there in the column Respondent? 

duplicate_count = df['Respondent'].duplicated().sum()

print(f"Number of duplicate values in the 'Respondent' column: {duplicate_count}")

Number of duplicate values in the 'Respondent' column: 154


## Removing duplicates


Remove the duplicate rows from the dataframe.


In [11]:
# your code goes here

# Remove duplicates
df_unique = df.drop_duplicates(subset = duplicate_rows)

Verify if duplicates were actually dropped.


In [12]:
# Calculate the number of rows before and after removing duplicates
rows_before = df.shape[0]
rows_after = df_unique.shape[0]

# Calculate the number of removed duplicates
removed_duplicates = rows_before - rows_after

# Print the results
print(f"Number of rows before removing duplicates: {rows_before}")
print(f"Number of rows after removing duplicates: {rows_after}")
print(f"Number of removed duplicates: {removed_duplicates}")
print(f"Number of duplicate rows: {duplicate_count}")

Number of rows before removing duplicates: 11552
Number of rows after removing duplicates: 11398
Number of removed duplicates: 154
Number of duplicate rows: 154


In [13]:
# additional code for graded quiz
# After removing the duplicate rows, how many blank rows are there under the column EdLevel?

# Count blank rows in the 'EdLevel' column
blank_edlevel_count = df_unique['EdLevel'].isna().sum()

print(f"Number of blank rows in the 'EdLevel' column after removing duplicates: {blank_edlevel_count}")

Number of blank rows in the 'EdLevel' column after removing duplicates: 112


## Finding Missing values


Find the missing values for all columns.


In [14]:
# your code goes here

# Set option to display all rows
pd.set_option('display.max_rows', None)

# Your existing code to calculate missing values
missing_values = df.isnull().sum()

# Display the missing values for all columns
print("Missing values for all columns:")
print(missing_values)


Missing values for all columns:
Respondent                   0
MainBranch                   0
Hobbyist                     0
OpenSourcer                  0
OpenSource                  81
Employment                   0
Country                      0
Student                     53
EdLevel                    116
UndergradMajor             740
EduOther                   164
OrgSize                     98
DevType                     67
YearsCode                    9
Age1stCode                  13
YearsCodePro                16
CareerSat                    0
JobSat                       1
MgrIdiot                   498
MgrMoney                   502
MgrWant                    498
JobSeek                      0
LastHireDate                 0
LastInt                    423
FizzBuzz                    37
JobFactors                   3
ResumeUpdate                41
CurrencySymbol               0
CurrencyDesc                 0
CompTotal                  815
CompFreq                   206
Convert

Find out how many rows are missing in the column 'WorkLoc'


In [25]:
# your code goes here

WorkLoc_missing_values = df['WorkLoc'].isnull().sum()
print(WorkLoc_missing_values)

32


## Imputing missing values


Find the  value counts for the column WorkLoc.


In [28]:
# your code goes here

# Find value counts for the WorkLoc column
work_loc_counts = df['WorkLoc'].value_counts()

# Display the results
print("Value counts for WorkLoc column:")
print(work_loc_counts)

Value counts for WorkLoc column:
WorkLoc
Office                                            6905
Home                                              3638
Other place, such as a coworking space or cafe     977
Name: count, dtype: int64


In [34]:
# Optional: Display as percentage

# Calculate percentages
work_loc_percentages = (work_loc_counts / len(df) * 100).round(2)

# Display the results
print("Value counts for WorkLoc column:")
print(work_loc_counts)

print("\nValue counts as percentage:")
for location, percentage in work_loc_percentages.items():
    print(f"{location}: {percentage:.2f}%")

Value counts for WorkLoc column:
WorkLoc
Office                                            6905
Home                                              3638
Other place, such as a coworking space or cafe     977
Name: count, dtype: int64

Value counts as percentage:
Office: 59.77%
Home: 31.49%
Other place, such as a coworking space or cafe: 8.46%


Identify the value that is most frequent (majority) in the WorkLoc column.


In [30]:
#make a note of the majority value here, for future reference

most_frequent_value = df['WorkLoc'].mode()[0]
print("\nMost frequent value in 'WorkLoc' column:")
print(most_frequent_value)


Most frequent value in 'WorkLoc' column:
Office


Impute (replace) all the empty rows in the column WorkLoc with the value that you have identified as majority.


In [39]:
# Impute the missing values in the 'WorkLoc' column with the most frequent value
df['WorkLoc'] = df['WorkLoc'].fillna(most_frequent_value)

After imputation there should ideally not be any empty rows in the WorkLoc column.


Verify if imputing was successful.


In [40]:
# your code goes here

# Verify that there are no remaining missing values in the 'WorkLoc'' column after imputation
remaining_missing_values = df['WorkLoc'].isnull().sum()

print("\nMost frequent value in 'WorkLoc'' column:")
print(most_frequent_value)

print("\nRemaining missing values in 'WorkLoc'' column after imputation:")
print(remaining_missing_values)


Most frequent value in 'WorkLoc'' column:
Office

Remaining missing values in 'WorkLoc'' column after imputation:
0


In [15]:
# additional code for graded quiz
# What is the majority category under the column Employment?

most_frequent_value2 = df['Employment'].mode()[0]
print(most_frequent_value2)

Employed full-time


In [16]:
# additional code for graded quiz
# Under the column " UndergradMajor", which category has the minimum number of rows?

# Find value counts for the WorkLoc column
UndergradMajor_counts = df['UndergradMajor'].value_counts()

# Calculate percentages
UndergradMajor_percentages = (UndergradMajor_counts / len(df) * 100).round(2)

# Display the results
print("Value counts for UndergradMajor column:")
print(UndergradMajor_counts)

print("\nValue counts as percentage:")
for major, percentage in UndergradMajor_percentages.items():
    print(f"{major}: {percentage:.2f}%")

Value counts for UndergradMajor column:
UndergradMajor
Computer science, computer engineering, or software engineering          7053
Information systems, information technology, or system administration     806
Another engineering discipline (ex. civil, electrical, mechanical)        769
Web development or web design                                             417
A natural science (ex. biology, chemistry, physics)                       409
Mathematics or statistics                                                 375
A business discipline (ex. accounting, finance, marketing)                247
A social science (ex. anthropology, psychology, political science)        212
A humanities discipline (ex. literature, history, philosophy)             209
Fine arts or performing arts (ex. graphic design, music, studio art)      165
I never declared a major                                                  126
A health science (ex. nursing, pharmacy, radiology)                        24
Name: cou

In [21]:
# additional code for graded quiz
# After removing the duplicate rows, how many respondents are being paid yearly?

# Remove duplicate rows
df_unique = df.drop_duplicates(subset=['Respondent'])

# Count respondents paid yearly
yearly_count = df_unique[df_unique['CompFreq'] == 'Yearly'].shape[0]

print(f"Number of respondents being paid yearly: {yearly_count}")


Number of respondents being paid yearly: 6073


## Normalizing data


There are two columns in the dataset that talk about compensation.

One is "CompFreq". This column shows how often a developer is paid (Yearly, Monthly, Weekly).

The other is "CompTotal". This column talks about how much the developer is paid per Year, Month, or Week depending upon his/her "CompFreq". 

This makes it difficult to compare the total compensation of the developers.

In this section you will create a new column called 'NormalizedAnnualCompensation' which contains the 'Annual Compensation' irrespective of the 'CompFreq'.

Once this column is ready, it makes comparison of salaries easy.


<hr>


List out the various categories in the column 'CompFreq'


In [18]:
# your code goes here

comp_freq_categories = df['CompFreq'].unique()
print("Categories in CompFreq column:")
print(comp_freq_categories)

Categories in CompFreq column:
['Yearly' 'Monthly' 'Weekly' nan]


Create a new column named 'NormalizedAnnualCompensation'. Use the hint given below if needed.


Double click to see the **Hint**.

<!--

Use the below logic to arrive at the values for the column NormalizedAnnualCompensation.

If the CompFreq is Yearly then use the exising value in CompTotal
If the CompFreq is Monthly then multiply the value in CompTotal with 12 (months in an year)
If the CompFreq is Weekly then multiply the value in CompTotal with 52 (weeks in an year)

-->


In [19]:
# your code goes here

import numpy as np

def normalize_annual_compensation(row):
    if pd.isna(row['CompFreq']) or pd.isna(row['CompTotal']):
        return np.nan
    elif row['CompFreq'] == 'Yearly':
        return row['CompTotal']
    elif row['CompFreq'] == 'Monthly':
        return row['CompTotal'] * 12
    elif row['CompFreq'] == 'Weekly':
        return row['CompTotal'] * 52
    else:
        return np.nan

# Create the new column
df['NormalizedAnnualCompensation'] = df.apply(normalize_annual_compensation, axis=1)

# Display the first few rows to verify the result
print(df[['CompFreq', 'CompTotal', 'NormalizedAnnualCompensation']].head(10))

# Display summary statistics
print("\nSummary statistics of NormalizedAnnualCompensation:")
print(df['NormalizedAnnualCompensation'].describe())

  CompFreq  CompTotal  NormalizedAnnualCompensation
0   Yearly    61000.0                       61000.0
1   Yearly   138000.0                      138000.0
2   Yearly    90000.0                       90000.0
3  Monthly    29000.0                      348000.0
4   Yearly    90000.0                       90000.0
5  Monthly     9500.0                      114000.0
6  Monthly     3000.0                       36000.0
7   Yearly   103000.0                      103000.0
8   Yearly    69000.0                       69000.0
9  Monthly     8000.0                       96000.0

Summary statistics of NormalizedAnnualCompensation:
count    1.073100e+04
mean     6.076494e+06
std      9.774169e+07
min      0.000000e+00
25%      5.200000e+04
50%      1.000000e+05
75%      3.600000e+05
max      8.400000e+09
Name: NormalizedAnnualCompensation, dtype: float64


In [20]:
df['NormalizedAnnualCompensation'].median()

100000.0

## Authors


Ramesh Sannareddy


### Other Contributors


Rav Ahuja


 ## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2024-09-24|1.1|Madhusudhan Moole|Updated lab|
|2024-09-23|1.0|Raghul Ramesh|Created lab|


<!--| Date (YYYY-MM-DD) | Version | Changed By        | Change Description                 |
| ----------------- | ------- | ----------------- | ---------------------------------- |
| 2020-10-17        | 0.1     | Ramesh Sannareddy | Created initial version of the lab |--!>


## <h3 align="center"> © IBM Corporation. All rights reserved. <h3/>
